BUGFIX: Serialised node fully qualified for fusion uncached mode

mhsdesign opened the pull request:
Resolves: #4732 

discussion should happen in #4732 and once we agree on a solution i will update this pr description

<!-- 
    Thanks for your contribution, we appreciate it!

    The first section should explain briefly what is changed. 
    Some examples are always nice to showcase the use. 
    The content will be used in the change-logs and addresses 
    developers working with Neos.

    If there are issues regarding the topic of your PR link 
    them here as `related:` or `resolved:`
-->

**Upgrade instructions**

<!-- 
    Add upgrade instructions for breaking changes. 
    Explain who is affected, what has to be adjusted  
-->

**Review instructions**

test that this works:

```
prototype(Neos.Demo:Content.Headline) < prototype(Neos.Neos:ContentComponent) {
    renderer = afx`...`

    @cache {
        mode = 'uncached'
        context {
            1 = 'node'
        }
    }
}
```

<!-- 
    If your change is not explained fully by the first section you can 
    add more details here to help the reviewers understand the change.
    We have to understand what you did, why you did it and how we can 
    verify it works correctly and does no harm.
-->

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

mhsdesign committed:
TASK: #4732 serialized node fully qualified for fusion uncached mode

kitsunet commented:
not a fan doing this specifically for fusion/the property mapper, why don't we replace the serialization format here `\Neos\Neos\FrontendRouting\NodeAddress::serializeForUri` with a fully qualified one?

kitsunet commented:
Or yes, we introduce NodeIdentity and drop the serialization format from NodeAddress, that would be even neater

-> https://github.com/neos/neos-development-collection/issues/4564

mhsdesign committed:
TASK: Refactor `RuntimeContentCache` serializeContext and unserializeContext to be in one place

mhsdesign commented:
Yes a `NodeIdentity` is the solution. And then we make use of json_serialise directly like the value object support for flow_json_array :D

mhsdesign committed:
TASK: Remove hard dependency from fusion cache serialization to flows property mapping

mhsdesign committed:
TASK: Introduce `NeosFusionContextSerializer` to replace node property converters

mhsdesign commented:
Solved it by removing the hardcoded property mapper from the fusion cache and using symfony's interfaces

kitsunet commented:
Wait what do we gain here? Another object? Like apart from abstracting the propertyMapper away here nothing has changed?

mhsdesign commented:
Oh this isnt just a new object.
This is so much more than an object :D

By adding a layer of abstraction Fusion standalone will use the property mapper yes, but Neos.Neos implements an own serialiser so that it can serialise nodes without bothering the propertymapper.
That allows us to remove property mapping for nodes e.g. [NewNodeConverter](https://github.com/neos/neos-development-collection/blob/9dd4ed5e657bcbb227b64e3991fc8268bdca539f/Neos.Neos/Classes/Fusion/NewNodeConverter.php#L36) and its college.
In my opinion its super important that people dont accidentally use the property mapper with the new nodes.
Allowing mapping would render all of our strictness and new explicitness useless as it would allow again magical nodes in controllers:
```
public function nodeAction(Node $node): {}
```
which will already now confusion for example about the nodes format when serialised ...

Lets to it rather explicit this time please :D

kitsunet commented:
> Allowing mapping would render all of our strictness and new explicitness useless as it would allow again magical nodes in controllers:

I think that is a core expectation though, countless projects will have controllers expecting a node and I don#t see why for the time being as the propertyMapper IS still the suggested and only way to do controller marshalling build into Flow, this should work

mhsdesign commented:
Huh i didnt know we werent on the same side :D I thought we discussed this already someplace but couldnt find it so i opened a separate issue: https://github.com/neos/neos-development-collection/issues/4873

Under that aspect, i will revert the removal of the property mappers with this pr.

mhsdesign commented:
We discussed that we dislike property mapping directly to a node via the flow property mapper. For fusion we have the usecase of unserialising the node `@cache.context { 0 = 'node' }`.
So injecting a dedicated specialised serialiser into the RuntimeContent cache should be the first step.
As followup as part of https://github.com/neos/neos-development-collection/issues/4564 we should unify the node serialisation across all places.

mhsdesign committed:
WIP: Re introduce hacky node converters, to be removed via #4873

mhsdesign committed:
TASK: Improve docs of `FusionContextSerializer`

mhsdesign commented:
This is now ready for review. Once we have a central NodeIdentity i will adjust the code to use that serialiser from the `NeosFusionContextSerializer`.
So please review the pr under that aspect and see if the general concepts make sense.

For example i decided against using a sympfony serializer wrapper for the normalizers, as its not configurable via objects yaml.
In the end this is a nice refactoring to centralise the serialisation to happen at one place (not RuntimeContentCache AND ContentCache)

Sebobo commented on Neos.Neos/Classes/Fusion/Cache/NeosFusionContextSerializer.php:
```diff
@@ -0,0 +1,130 @@
+
+        $node = $subgraph->findNodeById(NodeAggregateId::fromString($serializedNode['nodeAggregateId']));
+        if (!$node) {
+            throw new \RuntimeException(sprintf('Could not find serialized node %s.', json_encode($serializedNode)), 1699782153);
```
Hm, isn't this a probable case that a node was stored which doesn't exist anymore now?
Should we fail hard here, or do we have another way of handling this?

mhsdesign replied:
Yes we have no choice but to fail.

Alternatively we could just return `null`, which will most likely completely crash silently in the rendering something.

The idea is that if you use `@cache.context = { 0 = 'node' }` (which will write a cache entry into the content cache) and then delete the node, the whole fusion rendering cannot happen in the first place because the node is deleted.
Unless of course caching ... But cache flushing must cover this. Or a hard `cache:flushone Neos_Fusion_Content` 😅 

An interesting test case still ....

In 8.3 we also were not happy if the node could not be converted: https://github.com/neos/neos-development-collection/blob/072186e1601d82a8565942a05c9fe7253cf7f85d/Neos.ContentRepository/Classes/TypeConverter/NodeConverter.php#L166-L169

kitsunet replied:
That's not the same, in 8.3 this Error would result in the PropertyMapper returning null, thus mostly just breaking rejndering of the sub part here that needs the node. I think that's saner. You cannot control editor behavior...

mhsdesign replied:
thanks for the hint ^^ should we log anything in the silent fail case?

mhsdesign replied:
Aaaand adjusted to return null ;)  ~If the workspace doesnt exist, i will still throw as this is most likely a fundamental error and not an editor hickup... i think ...~ -> no ill also just return null.

kitsunet commented on Neos.Fusion/Classes/Core/Cache/RuntimeContentCache.php:
```diff
@@ -271,7 +274,7 @@ public function postProcess(array $evaluateContext, $fusionObject, $output)
                 $contextVariables = $contextArray;
             }
-            $output = $this->contentCache->createUncachedSegment($output, $evaluateContext['fusionPath'], $contextVariables);
+            $output = $this->contentCache->createUncachedSegment($output, $evaluateContext['fusionPath'], $this->serializeContext($contextVariables));
```
How is this supposed for anything that is not a node?

mhsdesign committed:
TASK: Dont throw error if node cannot be deserialized.

Sebobo reviewed (approved):
Fine by reading with my limited understanding of this specific topic.

mhsdesign committed:
TASK: Always fail silently in `tryDeserializeNode`

as this was the old Neos 8.3 behaviour when the property mapper was still used for nodes

kitsunet reviewed (approved):
Seems fine I guess, bit weird that the serializer creates an array but I guess that's legacy from how it was before

mhsdesign commented:
> creates an array

actually thats new :D As the content cache doesnt care as long as its json serialise able i decided for array.
And this is just an intermediate step, with https://github.com/neos/neos-development-collection/pull/4868 we can use the `NodeIdentity` dto here :D

Thats how the symfony denormaliser thing is allowed to work.
